### ISSUE TITLE
Java test runner on lasagna returns error on passing tests

### ISSUE THREAD

**Issue by pdmoore** [exercism/exercism#5806]
Handle: pdmoore
Track: java
Exercise: lasagna

Using CLI,  submitted a solution that passed all tests.
Refactored to extract constants, submitted, and Iteration 2 failed (message "In order for our systems to analyze your code, the tests must be passing."). Tests do pass on my local machine. Deleted this iteration.
Submitted same code as Iteration 3. Still failing.
Submitted same code as Iteration 4 and it passed.

**Comment by iHiD** [comment]
So here, the tests are running but some error is occurring in exiting of the test runner from what I can tell.

This is stdout:

```
^[[A=> "Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details\n> Task :buildSrc:compileJava UP-TO-DATE\n> Task :buildSrc:compileGroovy NO-SOURCE\n> Task :buildSrc:processResources NO-SOURCE\n> Task :buildSrc:classes UP-TO-DATE\n> Task :buildSrc:jar UP-TO-DATE\n> Task :buildSrc:assemble UP-TO-DATE\n> Task :buildSrc:compileTestJava NO-SOURCE\n> Task :buildSrc:compileTestGroovy NO-SOURCE\n> Task :buildSrc:processTestResources NO-SOURCE\n> Task :buildSrc:testClasses UP-TO-DATE\n> Task :buildSrc:test NO-SOURCE\n> Task :buildSrc:check UP-TO-DATE\n> Task :buildSrc:build UP-TO-DATE\n> Task :lasagna:addListeners\n> Task :compileJava NO-SOURCE\n> Task :lasagna:compileJava\n> Task :processResources NO-SOURCE\n> Task :classes UP-TO-DATE\n> Task :compileTestJava NO-SOURCE\n> Task :lasagna:processResources NO-SOURCE\n> Task :lasagna:classes\n> Task :lasagna:compileTestJava\n\nDeprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.\nUse '--warning-mode all' to show the individual deprecation warnings.\nSee https://docs.gradle.org/6.8.3/userguide/command_line_interface.html#sec:command_line_warnings\n\nBUILD SUCCESSFUL in 16s\n3 actionable tasks: 3 executed\nRunning tests\n> Task :buildSrc:compileJava UP-TO-DATE\n> Task :buildSrc:compileGroovy NO-SOURCE\n> Task :buildSrc:processResources NO-SOURCE\n> Task :buildSrc:classes UP-TO-DATE\n> Task :buildSrc:jar UP-TO-DATE\n> Task :buildSrc:assemble UP-TO-DATE\n> Task :buildSrc:compileTestJava NO-SOURCE\n> Task :buildSrc:compileTestGroovy NO-SOURCE\n> Task :buildSrc:processTestResources NO-SOURCE\n> Task :buildSrc:testClasses UP-TO-DATE\n> Task :buildSrc:test NO-SOURCE\n> Task :buildSrc:check UP-TO-DATE\n> Task :buildSrc:build UP-TO-DATE\n> Task :lasagna:addListeners\n> Task :compileJava NO-SOURCE\n> Task :processResources NO-SOURCE\n> Task :classes UP-TO-DATE\n> Task :compileTestJava NO-SOURCE\n> Task :processTestResources NO-SOURCE\n> Task :testClasses UP-TO-DATE\n> Task :test NO-SOURCE\n> Task :lasagna:compileJava UP-TO-DATE\n> Task :lasagna:processResources NO-SOURCE\n> Task :lasagna:classes UP-TO-DATE\n> Task :lasagna:compileTestJava UP-TO-DATE\n> Task :lasagna:processTestResources NO-SOURCE\n> Task :lasagna:testClasses UP-TO-DATE\n\n> Task :lasagna:test\n\nLasagnaTest > total_time_in_minutes_for_multiple_layers PASSED\n\nLasagnaTest > total_time_in_minutes_for_one_layer PASSED\n\nLasagnaTest > remaining_minutes_in_oven PASSED\n\nLasagnaTest > preparation_time_in_minutes_for_one_layer PASSED\n\nLasagnaTest > expected_minutes_in_oven PASSED\n\nLasagnaTest > preparation_time_in_minutes_for_multiple_layers PASSED\n\nDeprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.\nUse '--warning-mode all' to show the individual deprecation warnings.\nSee https://docs.gradle.org/6.8.3/userguide/command_line_interface.html#sec:command_line_warnings\n\nBUILD SUCCESSFUL in 2s\n4 actionable tasks: 2 executed, 2 up-to-date\n"
```

I'll look further into why we might be having an error exiting and see if I can debug.

(cc @jmrunkle)

**Comment by iHiD** [comment]
I've updated some code in the tooling orchestrator so we should get a better error message out again. Please can you report if you see another issue so I can look at the logs. Thanks.

**Comment by pdmoore** [comment]
I submitted to exercism again and it failed on that attempt - Iteration 5 of lasagna

**Comment by iHiD** [comment]
Yeah, that just timed out. It seems the test runner is just quite slow atm. Taking 16s to run anything (and the timeout is 10s). 

@jmrunkle Is there any way to speed this up. Would a bigger machine help?

**Comment by jmrunkle** [comment]
> @jmrunkle Is there any way to speed this up. Would a bigger machine help?

Possibly. I'm not as familiar with what exactly the test runner is doing but it seems to be running Gradle and then processing the results from Gradle. Java can definitely be somewhat slow and has fairly high overheads for the JVM compared to some leaner languages. I'm just surprised to see if be this bad for such simple pieces of code.

**Comment by iHiD** [comment]
It's taking 16s to compile the code from what I can see.

**Comment by jmrunkle** [comment]
It looks like we could use the --scan option on the Gradle calls to see where the time is being spent. I doubt using the parallel option would help since everything should be in one package / module.

Other options seem like we could abandon using Gradle for the test runner and try using something else (I see Ant or Maven recommended elsewhere). That said, it would be inconsistent with the way the rest of our stuff uses Gradle.

I am reviewing the java-test-runner repo to see if I can learn more about how it works to see if there are any low-hanging fruits.

**Comment by jmrunkle** [comment]
I see it has a significant number of dependencies to compile... maybe there's a way to cache those or have them locally stored? https://github.com/exercism/java-test-runner/blob/main/lib/buildSrc/build.gradle#L7-L15

**Comment by SleeplessByte** [comment]
Gradle cache can be added to the docker layer, then it could skip that. It should already do it because during test run there's no internet access.

**Comment by jmrunkle** [comment]
I'm not super docker literate. Anyone more familiar with that able to tell me if this looks right:

Dockerfile: https://github.com/exercism/java-test-runner/blob/main/Dockerfile
Actual runner Gradle calls are here: https://github.com/exercism/java-test-runner/blob/main/bin/run.sh

I'm gathering that the test runs rely on the data copied here: https://github.com/exercism/java-test-runner/blob/main/Dockerfile#L17

That being said, I think there is still code generation for example the lombok stuff here which could clog up some compile time: https://github.com/exercism/java-test-runner/blob/main/lib/buildSrc/src/main/java/com/exercism/data/Report.java

It might help if we saw more from an actual gradle `--scan` of the compilation to see where we are really losing time.

**Comment by ErikSchierboom** [comment]
> That being said, I think there is still code generation for example the lombok stuff here which could clog up some compile time: https://github.com/exercism/java-test-runner/blob/main/lib/buildSrc/src/main/java/com/exercism/data/Report.java

Is there a way we could pre-compile that?

> It might help if we saw more from an actual gradle --scan of the compilation to see where we are really losing time.

This is fairly easy to do with the java test runner's `run_in_docker.sh` script. For example, I can run `./run_in_docker.sh two-fer /home/erik/exercism/java/exercises/practice/two-fer /home/erik/exercism/java/exercises/practice/two-fer` and the test runner Docker image will be built and used to test the two-fer exercise.

I then tried to add a `gradle --scan` to the `bin/run.sh` file, but that lead to the following error:

```
Auto-applied by using --scan

* What went wrong:
Plugin [id: 'com.gradle.enterprise', version: '3.5', artifact: 'com.gradle:gradle-enterprise-gradle-plugin:3.5'] was not found in any of the following sources:

- Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
- Plugin Repositories (could not resolve plugin artifact 'com.gradle:gradle-enterprise-gradle-plugin:3.5')
  Searched in the following repositories:
    Gradle Central Plugin Repository
```

Any idea on how to fix this? And is running a `gradle --scan` command what you wanted?

**Comment by jmrunkle** [comment]
That is what I wanted but I do not know how to fix it. This is stretching my Gradle knowledge to its limits.

**Comment by msomji** [comment]
👋  I am just catching up on this issue, I am not seeing any errors running the test runner locally, however I did noticed that overall the test-runner is taking  more time to run each exercise with the `gradle:6.8.3-jdk11` docker image as opposed to the old one we were using earlier (`gradle:6.3.0-jdk8`). I am in the process of compiling a list of how much time it takes to run each exercise between the two images, (I have maxed out my docker pulls so have to wait 6 hours before I continue) but from a smaller sample size, I saw that the test-runner now takes about 5-9 seconds longer to run with the new image. 
i.e The two-fer exersice took ~12 secs using the older gradle image and now takes ~16 seconds.

Unfortunately I haven't been able to get the `gradle --scan` to work to specifically identify why the runner is taking so much longer. I did however notice that the `gradle:6.8.3-jdk11` image is ~100Mb larger than `gradle:6.3.0-jdk8`.

Would it be out of the question to excessively increase the timeouts? To like 30 seconds? The Docker build does cache the image, but compilation of the code is where I think we are taking a hit.

I also noticed that there were some deprecation warnings being thrown by gradle itself after the update in the gradle version- I'll create a PR for that so we don't see those as well.

**Comment by iHiD** [comment]
>Would it be out of the question to excessively increase the timeouts? To like 30 seconds? The Docker build does cache the image, but compilation of the code is where I think we are taking a hit.

In short, yeah. It would be out of the question. Mainly because someone could easily break our whole infrastructure by queuing a few iterations and then all the test-running machines would get locked. If/when we get this working on AWS Lambda, this would become more reasonable, but for now it wouldn't work.

We already currently have the timeouts set to 20s, which is 2x what we intend to have them at, just for this initial phase.

Is there any way to use a stripped down version of the libraries that Gradle is using. It seems very over-the-top that Java takes ~12secs to compile "Hello World" (I'm not disputing it, to be clear - it just is so much slower than most other languages), so I'm wondering if we could strip down what Gradle "does" somehow? Are there flags that might be valuable in production that aren't relevant here for example? e.g. Can we make runtime-performance hits to speed up the compilation bit? Are there any bits (e.g. the tests) that could be precompiled and cached to save time somehow?

**Comment by jmrunkle** [comment]
That is just the thing, since we have no dependencies outside of standard library that are allowed in the build, compilation to bytecode should be extremely fast. There is *something* that is taking up more time than it should during compilation. My initial suspect is code generation like lombok - when I was looking at the test runner repo, it looked like we were moving the students code into the same directory where the test runner code is and then running it there. I would have thought we could build the test runner code separately (and only once) so that we are only compiling the student's code (and the tests) while running tests. I could be totally wrong, that was just the way the code looked to me.

**Comment by jmrunkle** [comment]
NOTE: investigating this a bit more in-depth over in https://github.com/exercism/java-test-runner/issues/29

**Comment by masahx** [comment]
maybe you should put an option for rerun of tests for one uploaded iteration?

**Comment by ErikSchierboom** [comment]
@pdmoore Thanks to @andreaTP, the Java test runner should be faster now and no longer timeout. Could you retry running the tests?

**Comment by ErikSchierboom** [comment]
We've not had any more reports on this issue due to the Java test runner now being a lot faster, I'm closing this issue. Thanks everyone for helping solve this.
